Initial integration of Receptor into AWX

shanemcd opened the pull request.

shanemcd committed:
Initial EE integration

shanemcd committed:
Begin integrating receptor

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1482,38 +1479,85 @@ def run(self, pk, **kwargs):
-                                                           module_args,
-                                                           ident=str(self.instance.pk))
-                self.finished_callback(None)
+            if not isinstance(self.instance, ProjectUpdate):
```
This block is a little weird since they currently both appear to do the same thing.  Maybe it would be better to just leave an `if not isinstance(self.instance, ProjectUpdate):` statement w/ a code comment inside of it for now?

shanemcd replied:
I'm not quite ready for this level of feedback yet, hence the Draft status. This conditional is just a hack that I plan on removing.

ryanpetrello replied:
Okey-dokey.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1482,38 +1479,85 @@ def run(self, pk, **kwargs):
+            # to exit. A reference to the processor_future is passed into the cancel_watcher_future,
+            # Which exits if the job has finished normally. The context manager ensures we do not
+            # leave any threads laying around.
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
```
This is some fancy new python3 niceness, and I think it's a reasonable way to accomplish this, but we'll want to pay careful attention to integration results to make sure we don't notice any sort of weird race conditions/hangs related to our logging code and this threadpool.  This sort of thing has bitten us in the past due to cpython bugs e.g.,

https://bugs.python.org/issue37429
https://github.com/ansible/awx/commit/5364e7839724026
https://bugs.python.org/issue38884

shanemcd replied:
Would you feel better with using `ProcessPoolExecutor`?

ryanpetrello replied:
Nope (for reasons we privately discussed) 😄

ryanpetrello commented on execution-environment.yml:
```diff
@@ -2,9 +2,14 @@
   append:
     - COPY --from=quay.io/project-receptor/receptor /usr/bin/receptor /usr/bin/receptor
     - RUN mkdir -m 0770 /var/run/receptor
+    - RUN pip3 install remote-pdb
```
Did you intend to leave these in?

ryanpetrello replied:
@shanemcd do we still need this (and elsewhere in this PR)

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -1482,38 +1479,85 @@ def run(self, pk, **kwargs):
+            sockin.close()
+            sockout.close()
+
+            self.update_model(self.instance.pk, receptor_work_id=result['unitid'])
```
I was really hoping that receptor would let us specify the work ID, because it should be fairly trivial to give a unique ID with the fields already on the unified job model. Then we wouldn't have to muck with tracking a new field. However...

https://github.com/project-receptor/receptor/blob/f7c245c18911a7ea5edb808bf568c39be21b17a6/receptorctl/receptorctl/socket_interface.py#L148-L151

I realize that the receptorctl code shows no sign of supporting this. Need to cross-reference against the `work submit` CLI itself, and I'm a little rusty to get to its help text right now.

jbradberry replied:
cc @ghjm

shanemcd replied:
I dont think we want a receptor network being tied to a single Tower installation. If we made this value derived from something like the job id, we're going to run into conflicts.

AlanCoding commented on tools/docker-compose/receptor.conf:
```diff
@@ -0,0 +1,27 @@
+- work-kubernetes:
+    worktype: ocp
+    namespace: receptor
+    image: quay.io/shanemcd/ee
```
Ultimately this will need to come from the execution environment model?

That question might be getting too far ahead of things, and if it is, that's fine.

shanemcd commented:
Folks, please wait until this PR is marked as ready for review before piling on.

shanemcd committed:
Set SDB_NOTIFY_HOST for all processes

shanemcd committed:
Create receptor mesh in cluster development environment

shanemcd committed:
Actually cancel things

shanemcd committed:
Surface errors when launching jobs through Receptor

This will raise errors such as:

exec: "ansible-runner": executable file not found in $PATH

shanemcd committed:
Pull awx -> receptor job code into its own class

shanemcd committed:
FOLLOW UP ON THIS: Fix fact_cache directory location

The part where we pass in the runner params to the processor phase is
legit. Need to investigate why the fact_cache directory is no longer nested
under job.id.

shanemcd committed:
Update receptor config to allow for runtime options

shanemcd committed:
InstanceGroup#is_containerized -> InstanceGroup#is_container_group

shanemcd committed:
UnifiedJob#is_containerized -> UnifiedJob#is_container_group_task

shanemcd committed:
Deleted unused build_params_process_isolation method

shanemcd committed:
Delete some old container group v1 code

shanemcd committed:
Use official awx-ee by default

shanemcd committed:
Minimally functional container group v2 w/ receptor

shanemcd committed:
flake8

shanemcd committed:
Make receptorctl easier to use in dev env

shanemcd committed:
Dont zip/unzip private data dir for local jobs

shanemcd committed:
Create awx group in container

shanemcd committed:
Add PATH to blocked inventory source vars

This used to be skiped because PATH was already present in the env we
constructed for runner.

AlanCoding commented:
recheck

AlanCoding committed:
Use the fully qualified inventory plugin name only for foreman

shanemcd commented:
recheck

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+        # reading.
+        sockin, sockout = socket.socketpair()
+
+        threading.Thread(target=self.transmit, args=[sockin]).start()
```
You never do a `.join()` on this do you?

Maybe I'm going to far with this, but some thoughts:

 - it doesn't seem to make sense to start `self.processor` if the transmit thread is still running, so maybe a `.join()` on the transmit thread would be good to have assured before processing starts. Otherwise there's nothing to process.
 - We do need to run the cancel watcher while this is running, because it could get canceled while a large amount of data is being transmitted. Maybe we could poll the cancel callback until that transmit thread finishes?

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+        sockin.close()
+        sockout.close()
+
+        resultsock, resultfile = receptor_ctl.get_work_results(unit_id,
```
this is expecting the return from here right?

https://github.com/project-receptor/receptor/blob/e293394235082df0b5f2cf037a43d13c6a1f3787/receptorctl/receptorctl/socket_interface.py#L233

I'd like to make it a little more clear what `resultfile` is, which comes from `self._sockfile = self._socket.makefile('rwb')`

There's 2 types of things being passed around, sockets and sockfiles. Some are made here, and some come from receptor. I'd like to try to get the naming to be as indicative of this as possible, like `result_sockfile`

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+            self.runner_params['only_transmit_kwargs'] = True
+
+        ansible_runner.interface.run(streamer='transmit',
+                                     _output=_socket.makefile('wb'),
```
writable sockfile, in the same way that `resultfile` is. Am I getting this correct?

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+            if res.status == 'canceled':
+                receptor_ctl.simple_command(f"work cancel {unit_id}")
+                resultsock.shutdown(socket.SHUT_RDWR)
+                resultfile.close()
```
I don't understand why the shutdown/close of the socket here would be specific to canceling.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+
+    def cancel_watcher(self, processor_future):
+        while True:
+            if processor_future.done():
```
I'm getting more and more worried about the _transmit_ thread hitting an error while this cancel_watcher is serving as the main loop, and is being counted on for the exit condition.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+            first_future = concurrent.futures.wait(futures,
+                                                   return_when=concurrent.futures.FIRST_COMPLETED)
+
+            res = list(first_future.done)[0].result()
```
Just trying to follow the code and jot down notes - `[0]` here could give the result from either the processing thread or the cancel watcher thread. It just doesn't matter because they give the same format of output, and the cancel watcher may outright give the result from the processing thread.

AlanCoding committed:
Fix minor syntax error failing AdHocCommands

AlanCoding commented:
I want to test the scenario where the remote job is killed half-way through. I suspect that `ansible-runner process` may not be designed in a way that can realize that an exit. The only way I can see us dealing with that is adding a check for `receptorctl status` inside of the cancel callback loop - basically reaper logic inside the controller process.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+
+            receptor_params = {
+                "secret_kube_pod": spec_yaml,
+                "secret_kube_config": kubeconfig_yaml
```
there are some cases we are seeing:

```
Traceback (most recent call last):
  File "/home/ec2-user/tower-qa/tests/api/external_jobs/test_container_groups.py", line 313, in test_container_group_sliced_jobs_exceeding_resource_quota
    job.assert_status(['running', 'pending', 'waiting'])
  File "/home/ec2-user/venvs/venv/lib64/python3.6/site-packages/awxkit/api/mixins/has_status.py", line 93, in assert_status
    raise AssertionError(msg)
AssertionError: Job-3243 has status of error, which is not in ['running', 'pending', 'waiting'].
result_traceback:
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1425, in run
    res = receptor_job.run()
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 3036, in run
    params=self.receptor_params)
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/receptorctl/socket_interface.py", line 187, in submit_work
    raise RuntimeError(errmsg)
RuntimeError: Failed to start work unit: secret_kube_config provided but not allowed
```

Looks like 2 places? So not a lot, don't know what's up with that.

AlanCoding replied:
raised from

https://github.com/project-receptor/receptor/blob/e293394235082df0b5f2cf037a43d13c6a1f3787/receptorctl/receptorctl/socket_interface.py#L187

AlanCoding committed:
Fix obvious code error with foreman inventory

ryanpetrello committed:
introduce a new global flag for denoating K8S-based deployments

- In K8S-based installs, only container groups are intended to be used
  for playbook execution (JTs, adhoc, inventory updates), so in this
  scenario, other job types have a task impact of zero.
- In K8S-based installs, traditional instances have *zero* capacity
  (because they're only members of the control plane where services
  - http/s, local control plane execution - run)
- This commit also includes some changes that allow for the task manager
  to launch tasks with task_impact=0 on instances that have capacity=0
  (previously, an instance with zero capacity would never be selected
  as the "execution node"

This means that when IS_K8S=True, any Job Template associated with an
Instance Group will never actually go from pending -> running (because
there's no capacity - all playbooks must run through Container Groups).
For an improved ux, our intention is to introduce logic into the
operator install process such that the *default* group that's created at
install time is a *Container Group* that's configured to point at the
K8S cluster where awx itself is deployed.

jbradberry commented on awx/main/isolated/manager.py:
```diff
@@ -45,28 +44,12 @@ def __init__(self, event_handler, canceled_callback=None, check_callback=None, p
-                '{} ansible_ssh_user={}'.format(host, settings.AWX_ISOLATED_USERNAME)
-                for host in hosts
-            ])
+        inventory = '\n'.join([
```
a minor complaint, but `.join(...)` doesn't need a list comprehension

AlanCoding committed:
Exception handling to always release work units

AlanCoding reviewed (approved):
There are still 2 PRs open against this https://github.com/shanemcd/awx/pulls, but it's relatively easy to change the base.

There's a lot of work still going on, but with the latest round of issues and patches, I feel like the form of work is taking shape. This has been the primary branch we've worked off for a while, so we should go ahead and merge it into `execution-environments` whenever.

shanemcd committed:
Merge pull request #26 from shanemcd/ee-capacity-experiment

introduce a new global flag for denoating K8S-based deployments

shanemcd committed:
Merge pull request #25 from AlanCoding/release_handling

Exception handling to always release work units

shanemcd committed:
Pin to latest version of PyYAML

Fixes https://github.com/yaml/pyyaml/issues/478

shanemcd commented:
recheck

jakemcdermott commented:
![doit](https://user-images.githubusercontent.com/9753817/105897391-171faa00-5fe6-11eb-9363-ec1163d110d9.gif)
